contracts/src/usernames/UsernamesV1.sol

```diff
@@ -0,0 +1,152 @@
+            revert InvalidUsername();
+        }
+
+        _registerUsername(msg.sender, username, b);
+    }
+
+    function resignUsername() external {
+        string memory username = _usernames[msg.sender];
+        // If user already has a username
+        if (bytes(username).length > 0) {
+            _usernameExists[keccak256(bytes(_usernames[msg.sender]))] = false;
+            delete _usernames[msg.sender];
+
+            emit UsernameResigned(msg.sender, username);
+        } else {
+            revert UsernameNotRegistered();
+        }
+    }
+
+    // External functions that are view
+    function version() external pure returns (uint256) {
+        return 1;
+    }
+
+    function getUsername(address user) external view returns (string memory) {
+        return _usernames[user];
+    }
+
+    function isUsernameRegistered(string memory username) external view returns (bool) {
+        return _usernameExists[keccak256(bytes(username))];
+    }
+
+    function isUsernameValid(string memory username) external pure returns (bool) {
+        return _verifyUsername(bytes(username));
+    }
+
+    // Internal function
+    // RULES:
+    // minimum length of 1 character
+    // maximum length of 20 characters
+    // only lowercase letters, numbers and underscores are allowed
+    // cannot start or end with underscore
+    // cannot contain two or more consecutive underscores
+    function _verifyUsername(bytes memory username) internal pure returns (bool) {
+        // minimum length of 1 character
+        // maximum length of 20 characters
+        if (username.length < 1 || username.length > 20) {
+            return false;
+        }
+
+        // cannot start or end with underscore
+        if (username[0] == 0x5F || username[username.length - 1] == 0x5F) {
+            return false;
+        }
+
+        for (uint256 i = 0; i < username.length; i++) {
+            // only lowercase letters, numbers and underscores are allowed
+            if (
+                !(username[i] >= 0x30 && username[i] <= 0x39) // 0-9
+                    && !(username[i] >= 0x61 && username[i] <= 0x7A) // a-z
+                    && !(username[i] == 0x5F) // _
+            ) {
+                return false;
+            }
+
+            // No need to care out ot bound access at i + 1, because previous test already check that latest character is not underscore
+            if (username[i] == 0x5F && username[i + 1] == 0x5F) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    function _registerUsername(address user, string memory username, bytes memory b) internal {
+        bytes32 usernameHash = keccak256(b);
+        if (_usernameExists[usernameHash]) {
+            revert TakenUsername();
+        }
+
+        // If user already has a username
+        if (bytes(_usernames[user]).length > 0) {
+            _usernameExists[keccak256(bytes(_usernames[user]))] = false; // Remove old username
```

Comment:
We could add an event when this happens (e.g. `UsernameUpdated` or `UsernameReplaced`) or simply include the old username in the `UsernameRegistered` event.